compiler/rustc_borrowck/src/diagnostics/conflict_errors.rs

```diff
@@ -1229,6 +1239,131 @@ impl<'cx, 'tcx> MirBorrowckCtxt<'cx, 'tcx> {
+    /// ```
+    fn suggest_using_closure_argument_instead_of_capture(
+        &self,
+        err: &mut Diagnostic,
+        borrowed_place: Place<'tcx>,
+        issued_spans: &UseSpans<'tcx>,
+    ) {
+        let &UseSpans::ClosureUse { capture_kind_span, .. } = issued_spans else { return };
+        let tcx = self.infcx.tcx;
+        let hir = tcx.hir();
+
+        // Get the type of the local that we are trying to borrow
+        let local = borrowed_place.local;
+        let Some(local_ty) = self
+            .body
+            .local_decls
+            .get(local)
+            .map(|l| l.ty)
+            else { return };
+
+        // Get the body the error happens in
+        let &body_id =
+            if let hir::Node::Item(hir::Item { kind, .. }) = hir.get(self.mir_hir_id())
+                && let hir::ItemKind::Static(_, _, body_id)
+                     | hir::ItemKind::Const(_, body_id)
+                     | hir::ItemKind::Fn(_, _, body_id) = kind
+            {
+                body_id
+            } else if let hir::Node::TraitItem(hir::TraitItem { kind, .. }) = hir.get(self.mir_hir_id())
+                && let hir::TraitItemKind::Const(_, Some(body_id))
+                     | hir::TraitItemKind::Fn(_, hir::TraitFn::Provided(body_id)) = kind
+            {
+                body_id
+            }else if let hir::Node::ImplItem(hir::ImplItem { kind, .. }) = hir.get(self.mir_hir_id())
+                && let hir::ImplItemKind::Const(_, body_id)
+                     | hir::ImplItemKind::Fn(_, body_id) = kind
+            {
+                body_id
+            } else {
+                return
+            };
+
+        let body_expr = hir.body(body_id).value;
+
+        struct ClosureFinder<'hir> {
+            borrow_span: Span,
+            res: Option<(&'hir hir::Expr<'hir>, &'hir hir::Closure<'hir>)>,
+        }
+        impl<'hir> Visitor<'hir> for ClosureFinder<'hir> {
+            fn visit_expr(&mut self, ex: &'hir hir::Expr<'hir>) {
+                if let hir::ExprKind::Closure(closure) = ex.kind
+                    && ex.span.contains(self.borrow_span)
+                    // To support cases like `|| { v.call(|this| v.get()) }`
+                    // FIXME: actually support such cases (need to figure out how to move from the capture place to original local)
+                    && self.res.as_ref().map_or(true, |(prev_res, _)| prev_res.span.contains(ex.span))
+                {
+                    self.res = Some((ex, closure));
+                }
+
+                hir::intravisit::walk_expr(self, ex);
+            }
+        }
+
+        // Find the closure that most tightly wraps `capture_kind_span`
+        let mut finder = ClosureFinder { borrow_span: capture_kind_span, res: None };
+        finder.visit_expr(body_expr);
+        let Some((closure_expr, closure)) = finder.res else { return };
+
+        let typeck_results: &TypeckResults<'_> = tcx.typeck(self.mir_def_id());
```

Comment:
I have no idea what this is 😅

Reply:
`ty::WithOptConstParam` is the device we use to avoid cycle errors for constants used as arguments for const generics.

```suggestion
        let typeck_results: &TypeckResults<'_> = tcx.typeck_opt_const_arg(self.body.source.with_opt_param());
```